xray/core/dataarray.py

```diff
@@ -1189,6 +1189,87 @@ def diff(self, dim, n=1, label='upper'):
         ds = self._dataset.diff(n=n, dim=dim, label=label)
         return self._with_replaced_dataset(ds)
 
+    def average(self, dim=None, axis=None, weights=None, returned=False,
+                keep_attrs=False, **kwargs):
+        """
+        Reduce this DataArray's data by applying `average` along some
+        dimension(s).
+
+        Parameters
+        ----------
+        dim : str or sequence of str, optional
+            Dimension(s) over which to apply `average`.
+        axis : int or sequence of int, optional
+            Axis(es) over which to apply `average`. Only one of the 'dim'
+            and 'axis' arguments can be supplied. If neither are supplied, then
+            `average` is calculated over axes.
+        weights : DataArray, optional
+            An array of weights associated with the values in this DataArray.
+            Each value in a contributes to the average according to its
+            associated weight. The weights array can either be 1-D (in which
+            case its length must be the size of a along the given axis or
+            dimension) or of he same shape this DataArray. If weights=None, then
+            all data in this DataArray are assumed to have a weight equal to one.
+        keep_attrs : bool, optional
+            If True, the attributes (`attrs`) will be copied from the original
+            object to the new one.  If False (default), the new object will be
+            returned without attributes.
+        returned : bool, optional
+            Default is False. If True, the tuple (average, sum_of_weights) is
+            returned, otherwise only the average is returned. If weights=None,
+            sum_of_weights is equivalent to the number of elements over which
+            the average is taken.
+
+        **kwargs : dict
+            Additional keyword arguments passed on to `sum`.
+
+        Returns
+        -------
+        reduced, [sum_of_weights] : DataArray, [DataArray]
+            New DataArray object with `average` applied to its data and the
+            indicated dimension(s) removed. When returned is True, return a
+            tuple with the average as the first element and the sum of the
+            weights as the second element. The return type is Float if a is of
+            integer type, otherwise it is of the same type as a.
+            sum_of_weights is of the same type as average.
+
+        See Also
+        --------
+        Dataset.average
+        DataArray.mean
+        """
+
+        if weights is None:
+            mean = self.mean(dim=dim, axis=axis, keep_attrs=keep_attrs)
+            if not returned:
+                return mean
+            else:
+                return (mean, self.size)
+        elif not isinstance(weights, DataArray):
+            # TODO: coerce weights of other types to dataarray?
```

Comment:
I think we need to do some refactoring to pull the logic from `_binary_op` out if we want to do this right. I was able to use `_binary_op` for `fillna` and `where` with [some hacks]9ttps://github.com/xray/xray/blob/5109f4f7bcb08c03ba006d42bed906ff0d194a29/xray/core/ops.py#L422-L429) but I don't think that will work here. The difficulty is that this a binary operation that also does a reduce.

Ideally, we would have some generic wrapping machinery that adds broadcasting and alignment for an arbitrary number of argument. For example, given a core signature like `'...a,....a->...'`  (using notation like [einsum](http://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum.html)) and a function defined on dask/numpy arrays, return an xray aware function that does alignment, broadcasting, looping over dataset variables and coordinate merging.

This would be useful in xray, but also in other projects that build on it -- for example, we have a library implementing [proper scoring rules](https://github.com/TheClimateCorporation/properscoring) that we just open sourced that has several function that would be useful to define on xray objects, with signatures like `'...,...e->...'` or `'...,...e,t->...t'`.